src/NuGet.Clients/NuGet.SolutionRestoreManager/ErrorListTableDataSource.cs

```diff
@@ -0,0 +1,176 @@
+﻿using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.TableControl;
+using Microsoft.VisualStudio.Shell.TableManager;
+using NuGet.PackageManagement.UI;
+
+namespace NuGet.SolutionRestoreManager
+{
+    /// <summary>
+    /// Add/Remove warnings/errors from the error list.
+    /// This persists messages once they are added.
+    /// </summary>
+    [Export(typeof(ErrorListTableDataSource))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal sealed class ErrorListTableDataSource : ITableDataSource
+    {
+        private readonly object _lockObj = new object();
+        private readonly IServiceProvider _serviceProvider;
+
+        private TableSubscription _tableSubscription;
+
+        public string SourceTypeIdentifier => StandardTableDataSources.ErrorTableDataSource;
+
+        public string Identifier => "NuGetRestoreManagerListTable";
+
+        public string DisplayName => "NuGet_Restore_Manager_Table_Data_Source";
+
+        public IErrorList _errorList;
+        public ITableManager _tableManager;
+        private bool _initialized;
+
+        [ImportingConstructor]
+        public ErrorListTableDataSource(
+            [Import(typeof(SVsServiceProvider))]
+            IServiceProvider serviceProvider)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            _serviceProvider = serviceProvider;
+        }
+
+        public IDisposable Subscribe(ITableDataSink sink)
+        {
+            _tableSubscription = new TableSubscription(sink);
```

Comment:
there can be multiple sink per data source.  you probably need a list of subscription.